petastorm/pytorch.py

```diff
@@ -423,3 +425,160 @@ def __enter__(self):
     def __exit__(self, exc_type, exc_val, exc_tb):
         self.reader.stop()
         self.reader.join()
+
+
+class AsyncBatchedDataLoader(BatchedDataLoader):
```

Comment:
It's a matter of personal style, but I'd typically avoid functional inheritance when not really-really appropriate. The main reason is that the interface between inherited and the parent object is not easy to track. For example, it took me some time to understand which methods of the parent are being called and which methods are reimplemented in this class.

I wonder, how would the code look like, if the reusable functions would be just free functions? E.g. we could define a free function
```python
def _yield_batches(buffer, keys) :
```
that could be used by both `BatchedDataLoader` and `AsyncBatchedDataLoader` even if we drop the inheritance relation.

Would also need to make `_instantiate_buffers` and `_iter_impl_worker` free functions. 

I won't insist on this design, but curious what are your thoughts on this?

Reply:
This seems very neat, I will think about it.